packages/nimble/inst/tests/test-macros.R

```diff
@@ -0,0 +1,222 @@
+source(system.file(file.path('tests', 'test_utils.R'), package = 'nimble'))
+
+nimbleOptions(enableModelMacros = TRUE)
+
+context("Testing model macros")
+
+test_that('Macro expansion 1',
+{
+    ## This converts a ~ testMacro1(b)
+    ## to inputs as (stoch = TRUE, LHS = a, b)
+    testMacro1 <- nimble:::model_macro_builder(
+        function(stoch, LHS, RHSarg) {
+            ans <- substitute(
+                OP(LHS, dnorm(RHSarg, 1)),
+                list(LHS = LHS, RHSarg = RHSarg,
+                     OP = if(stoch) as.name('~') else as.name('<-'))
+            )
+            list(code = ans)
+        },
+        use3pieces = TRUE, ## default
+        unpackArgs = TRUE  ## default
+    )
+    temporarilyAssignInGlobalEnv(testMacro1)
+    
+    expect_identical(
+        nimble:::codeProcessModelMacros(
+            quote(x[1] ~ testMacro1(y[1]))
+        ),
+        quote(x[1] ~ dnorm(y[1], 1))
+    )
+    
+    model1 <- nimbleModel(
+        nimbleCode({
+            x[1] ~ testMacro1(y[1])
+        })
+    )
+    expect_identical(
+        model1$getCode()[[2]],
+        quote(x[1] ~ dnorm(y[1], 1))
+    )
+
+    model2 <- nimbleModel(
+        nimbleCode({
+            for(i in 1:3)
+                x[i] ~ testMacro1(y[i])
+            for(j in 1:3) {
+                a[j] ~ testMacro1(b[j])
+            }
+            for(i in 1:3)
+                for(j in 1:3)
+                    f[i,j] ~ testMacro1(g[i])
+        })
+    )
+    expandedCode <- model2$getCode()
+    expect_identical(
+        expandedCode[[2]][[4]],
+        quote(x[i] ~ dnorm(y[i], 1))
+    )
+    expect_identical(
+        expandedCode[[3]][[4]][[2]],
+        quote(a[j] ~ dnorm(b[j], 1))
+    )
+    expect_identical(
+        expandedCode[[4]][[4]][[4]],
+        quote(f[i, j] ~ dnorm(g[i], 1))
+    ) 
+})
+
+
+test_that('Macro expansion 2',
+{
+    ## Like testMacro1 but with unpackArgs = FALSE.
+    ## This takes a ~ testMacro2(b) with arguments (stoch = TRUE, LHS = a, RHS = b)
+    testMacro2 <- nimble:::model_macro_builder(
+        function(stoch, LHS, RHS) {
+            if(RHS[[1]] != "testMacro2") stop("Problem with how testMacro2 was called")
+            RHS[[1]] <- as.name('step') ## something valid
+            ans <- substitute(
+                OP(LHS, dnorm(RHS, 1)),
+                list(LHS = LHS, RHS = RHS,
+                     OP = if(stoch) as.name('~') else as.name('<-'))
+            )
+            list(code = ans)
+        },
+        use3pieces = TRUE, ## default
+        unpackArgs = FALSE  ## NON-default
+    )
+    
+    temporarilyAssignInGlobalEnv(testMacro2)
```

Comment:
Do I understand correctly that the `temporarilyAssignInGlobalEnv` is required by `nimbleModel` but not by `nimble:::codeProcessModelMacros`, since the latter searches in its parent frame? If that is the case, it might be a good global fix to modify `nimbleModel` to search in its parent frame as well. Do you think that would work?